src/osd/OSD.h

```diff
@@ -1745,67 +1646,98 @@ class OSD : public Dispatcher,
   const unsigned int op_prio_cutoff;
 
   friend class PGQueueable;
-  class ShardedOpWQ: public ShardedThreadPool::ShardedWQ < pair <PGRef, PGQueueable> > {
-
+  class ShardedOpWQ : public ShardedThreadPool::ShardedWQ<pair<spg_t,PGQueueable>> {
     struct ShardData {
       Mutex sdata_lock;
       Cond sdata_cond;
       Mutex sdata_op_ordering_lock;
-      map<PG*, list<PGQueueable> > pg_for_processing;
-      std::unique_ptr<OpQueue< pair<PGRef, PGQueueable>, entity_inst_t>> pqueue;
+
+      OSDMapRef waiting_for_pg_osdmap;
+      struct pg_slot {
+	PGRef pg;
+	list<PGQueueable> ls;
+	bool waiting_for_pg = false;
+	uint64_t requeue_seq = 0;
+      };
+      unordered_map<spg_t,pg_slot> pg_slots;
+
+      std::unique_ptr<OpQueue< pair<spg_t, PGQueueable>, entity_inst_t>> pqueue;
+
+      void _enqueue_front(pair<spg_t, PGQueueable> item, unsigned cutoff) {
```

Comment:
we just don't have a local pointer back to osd in ShardData.  an arg here was cleaner than adding an extra field